src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java

```diff
@@ -1,60 +1,152 @@
 import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.apache.commons.lang3.StringUtils.SPACE;
+import static org.apache.commons.lang3.StringUtils.containsNone;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static org.apache.commons.lang3.StringUtils.isNotEmpty;
+
 
 /**
- * Example of ApplicationContext implementation, can be removed or reimplemented when actual development starts
+ * Implementation of ApplicationContext
  */
+@Slf4j
 public class ApplicationContextImpl implements ApplicationContext {
+
+    private Table<String, Class<?>, BeanDefinition> beanDefinitionTable;
+
+    /**
+     * Scanners scan application packages to define {@link BeanDefinition} configs.
+     * Received list of bean definitions validates by validator to avoid duplicates and other problems (see {@link BeanValidator}).
+     * Beans definitions list maps to container structure. It's Table<String, Class<?>, BeanDefinition>, where
+     * - row key is a name of bean
+     * - column key is a type of bean
+     * - mapped value is bean definition
+     * Then context passes to {@link BeanInitializer} for beans instances initialization
+     *
+     * @param scanners    list of scanners for
+     * @param validator   bean definition validator
+     * @param initializer bean initializer
+     */
     public ApplicationContextImpl(List<BeanScanner> scanners,
                                   BeanValidator validator,
                                   BeanInitializer initializer) {
-        /* 1) Scanners get all beans
-           2) Merge List of bean definitions into one List
-           3) Send beans list to validator to check duplicates
-           4) Convert bean list to map
-           5) Send beans map to initializer to instantiate beans and inject dependencies
-         */
+
+        List<BeanDefinition> beanDefinitionList = runScanning(scanners);
+        validator.validate(beanDefinitionList);
+        beanDefinitionTable = populateBeansDefinitionTable(beanDefinitionList);
+        initializer.initialize(beanDefinitionTable);
+
+        log.info("Application context initialized");
     }
 
-    @Override
-    public <T> void register(String name, Class<T> beanType) {
+    private List<BeanDefinition> runScanning(List<BeanScanner> scanners) {
+        log.info("Scanning application for finding bean definitions");
+
+        return scanners.stream()
+                .map(BeanScanner::scan)
+                .flatMap(List::stream)
+                .toList();
+    }
 
+    private Table<String, Class<?>, BeanDefinition> populateBeansDefinitionTable(List<BeanDefinition> beanDefinitionList) {
+        log.debug("Store scanned bean definition in application context");
+        HashBasedTable<String, Class<?>, BeanDefinition> beanDefinitionTable = HashBasedTable.create();
```

Comment:
You should use either polymorphic reference `Table` or use `var` instead for simplicity